pkg/controllers/node/node_controller.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"reflect"
+	"sync"
+
+	"time"
+
+	calicocache "github.com/projectcalico/k8s-policy/pkg/cache"
+	"github.com/projectcalico/k8s-policy/pkg/controllers/controller"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/fields"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	corecache "k8s.io/client-go/tools/cache"
+
+	"fmt"
+)
+
+// This cache contains kubernetes node information keyed by etcd node-name.
+type cache struct {
+	sync.RWMutex
+	// The key for the cache's data is an etcd nodeName.
+	data map[string]string
+}
+
+// NodeController implements the Controller interface.  It is responsible for monitoring
+// kubernetes nodes and responding to delete events by removing them from the Calico datastore.
+// It keeps a cache of known calico nodes and their corresponding kubernetes nodes to
+// accomplish this.
+type NodeController struct {
+	indexer       corecache.Indexer
+	informer      corecache.Controller
+	nodeDataCache calicocache.ResourceCache
```

Comment:
For clarity, I'd suggest renaming the two caches:

`kubernetesResourceCache` for one and something like `workerNodeCache` or `calicoNodeCache` for the other.